src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs

```diff
@@ -0,0 +1,293 @@
+            {
+                Initialize(block, isStartBlock);
+            }
+            else
+            {
+                ClearPendingValidatorsAfterReorganisation(block);                
+            }
+
+            FinalizePendingValidatorsIfNeeded(block);
+        }
+
+        private void ClearPendingValidatorsAfterReorganisation(Block block)
+        {
+            bool reorganisationHappened = block.Number <= _validatedSinceFinalized.LastBlockNumber;
+            if (reorganisationHappened)
+            {
+                bool reorganisationBeforeInitChange = block.Number <= _pendingValidators.BlockNumber;
+                if (reorganisationBeforeInitChange)
+                {
+                    ClearPendingValidators();
+                }
+                else
+                {
+                    _validatedSinceFinalized.RemoveFromBlock(block.Number);
+                }
+            }
+        }
+
+        private void FinalizePendingValidatorsIfNeeded(Block block)
+        {
+            bool thereArePendingValidators = _pendingValidators != null;
+            if (thereArePendingValidators)
+            {
+                var shouldFinalizePendingValidators = _validatedSinceFinalized.Count > 0 && _validatedSinceFinalized.Count > _validators.Count / 2;
+                if (shouldFinalizePendingValidators)
+                {
+                    if (_logger.IsInfo) _logger.Info($"Applying validator set change signalled at block {_pendingValidators.BlockNumber} at block {block.Number}.");
+                    ValidatorContract.InvokeTransaction(block.Header, _transactionProcessor, ValidatorContract.FinalizeChange(), Output);
+                    
+                    _validators = new HashSet<Address>(_pendingValidators.Addresses);
+                    ClearPendingValidators();
+                }
+                else
+                {
+                    _validatedSinceFinalized.Add(new ValidationStamp(block));                    
+                }
+            }
+        }
+
+        public virtual void PostProcess(Block block, TxReceipt[] receipts)
+        {
+            if (ValidatorContract.CheckInitiateChangeEvent(ContractAddress, block, receipts, out var potentialValidators))
+            {
+                InitiateChange(block, potentialValidators);
+                if (_logger.IsInfo) _logger.Info($"Signal for transition within contract at block {block.Number}. New list: [{string.Join<Address>(", ", potentialValidators)}].");
+            }
+        }
+
+        public virtual AuRaParameters.ValidatorType Type => AuRaParameters.ValidatorType.Contract;
+        
+        protected virtual ValidatorContract CreateValidatorContract(Address contractAddress) => 
+            new ValidatorContract(AbiEncoder, contractAddress);
+
+        private void Initialize(Block block, bool isStartBlock)
+        {
+            if (isStartBlock)
+            {
+                CreateSystemAccount();
```

Comment:
I think that SystemAccount will be more general than just a contract validator related account

Reply:
Maybe yes, still it is created at first usage on block 1 of using contract validator on AuRa networks (not block 0). Theoretically we might have a fallback and create it during processing. This can be discussed.